### PR TITLE
Nix include git untracked files

### PR DESCRIPTION
**Pull request by jfroche**
Patch Nix so that it doesn't ignore untracked files in git anymore and
continue to ignore files if they are in a .gitignore.

### CODE REVIEW

File: src/Garn.hs [hunk 1]
```diff
@@ -88,4 +88,4 @@ checkTarget targetConfig = case targetConfig of
   TargetConfigExecutable _ -> pure ()
 
 runNix :: (Input a, Output b) => a -> IO b
-runNix = Cradle.run "nix" nixArgs
+runNix = Cradle.run "nix" "--include-untracked-files" nixArgs
```
Comment (soenkehahn) [inline src/Garn.hs:91]: I think this should go into `nixArgs`.

File: flake.nix [hunk 2]
```diff
@@ -15,6 +15,25 @@
...
+                    (final.fetchpatch {
+                      name = "include-untracked-files.patch";
+                      url = "https://github.com/NixOS/nix/pull/9352.patch";
+                      excludes = [ "tests/functional/*" ];
+                      hash = "sha256-JPoQSXFKex6wFx2AoOuMTY0PvzdZ9R5wCqcsT+13i7k=";
+                    })
+                  ];
+                  # Set big-parallel, this allows us to build nix remotely
+                  # on machines where only big-parallel derivations are built
+                  # remotely.
```
Comment (soenkehahn) [inline flake.nix:32]: I'm tempted to delete this comment. If a reader has no idea what `big-parallel` is, this comment is not that helpful without actually googling and reading up on `big-parallel`. But for that, we don't need a comment. WDYT?

**Review by soenkehahn — COMMENTED** [review]
I think we should have a test (maybe in `GarnSpec.hs`) that `garn` behaves this way now. Happy to pair on that.

File: flake.nix [hunk 3]
```diff
@@ -15,6 +15,22 @@
...
           inherit system;
           config.allowUnfree = true;
+          overlays = [
+            (final: prev: {
+              nix = prev.nix.overrideAttrs
+                (finalAttrs: prevAttrs: {
+                  patches = (prevAttrs.patches or [ ]) ++ [
+                    (final.fetchpatch {
+                      name = "include-untracked-files.patch";
+                      url = "https://github.com/NixOS/nix/compare/2.17.1...jfroche:nix:feat/include-untracked-files-2.17.1.patch";
```
Comment (jfroche) [inline flake.nix:25]: I have rebased the pull request in the Nix repository. Due to the recent changes in the main branch, specifically the transition to the libgit2 git implementation, the PR can no longer be directly applied with our current Nix version (2.17.1).
To address this, I've initiated a branch to backport the necessary changes for version 2.17.1.

**Review by alexdavid — APPROVED** [review]
This looks good to me. I think we can also remove the `git add --intent-to-add flake.nix` call in `src/Garn/GarnConfig.hs:192` as well as the FAQ entry on the website at `website/src/pages/main.tsx:263-281`

**Comment by jfroche** [comment]
We still get the output of the `git init` during the test, with something like this:

```
    builds packages in a git repository with untracked files [ ]Initialized empty Git repository in /tmp/nix-shell.P8XVK8/mockery-5db85674c541bbc8/.git/
    builds packages in a git repository with untracked files [✔]
 ```
 Not sure what would be the best to remove that.
 
I also see a lot of `warning: Git tree '...' is dirty` by nix during the tests, would it make sense to run nix with `--no-warn-dirty` to maybe reduce confusion for new comers ?

File: test/spec/BuildSpec.hs [hunk 4]
```diff
@@ -190,3 +190,11 @@ spec = do
           output <- runGarn ["build", "project"]
           readFile "result/build-artifact" `shouldReturn` "hello from setup\n"
           exitCode output `shouldBe` ExitSuccess
+
+      it "builds packages in a git repository with untracked files" $ \runGarn -> do
```
Comment (soenkehahn) [inline test/spec/BuildSpec.hs:194]:
```suggestion
      it "includes untracked files when building packages" $ \runGarn -> do
```

**Review by soenkehahn — APPROVED** [review]

**Pull request merged**